src/main/kotlin/compiler/Coder.kt

```diff
@@ -99,4 +99,90 @@ class Coder(val ast: Node) {
         }
         return s
     }
+
+
+    object Optimizer {
+        var source = ArrayList<VMWord>()
+        var mem = ArrayList<VMWord>()
+        val jumpMap = HashMap<Int, Int>()
+        var pc = 0
+        var lastMatchSize = 0
+
+        fun postOptimize(withSource: ArrayList<VMWord>) {
+
+            // Find all jump destinations in source
+            source = withSource
+            source.forEach { word ->
+                word.address?.also { address ->
+                    jumpMap[address] = -1
+                }
+            }
+
+            mem.clear()
+            pc = 0
+            while (pc < source.size) {
+                // If we've reached a jump dest, record its new address
+                if (jumpMap.containsKey(pc)) jumpMap[pc] = mem.size
+
+
+                // NEGATE NEGATE => ()
+                consume(O_NEGATE, O_NEGATE)?.also { }
+
+                // SETVAR GETVAR => SETGETVAR
+                ?: consume(O_SETVAR, null, O_GETVAR, null) { args ->
+                    args[0].isInt() && args[1].isInt(args[0].intFromV)
+                }?.also { args ->
+                    code(O_SETGETVAR)
+                    value(args[0].value!!)
+                }
+
+                // O_VAL 0 O_CMP_xx => O_CMP_xxZ
+                ?: consume(O_VAL, null, O_CMP_EQ) { args -> args[0].isInt(0) }?.also { code(O_CMP_EQZ) }
+                ?: consume(O_VAL, null, O_CMP_GT) { args -> args[0].isInt(0) }?.also { code(O_CMP_GTZ) }
+                ?: consume(O_VAL, null, O_CMP_GE) { args -> args[0].isInt(0) }?.also { code(O_CMP_GEZ) }
+                ?: consume(O_VAL, null, O_CMP_LT) { args -> args[0].isInt(0) }?.also { code(O_CMP_LTZ) }
+                ?: consume(O_VAL, null, O_CMP_LE) { args -> args[0].isInt(0) }?.also { code(O_CMP_LEZ) }
+
+
+                // If nothing matched, copy and continue
+                ?: run {
+                    mem.add(source[pc++])
+                }
+            }
+
+            // Replace all jump dests
+            jumpMap.keys.forEach { old ->
+                mem.forEach { word ->
+                    if (word.address == old) word.address = jumpMap[old]
+                }
+            }
+        }
+
+        // Match and consume a series of opcodes (or null for any non-opcode word).
+        private fun consume(vararg opcodes: Opcode?, check: ((List<VMWord>)->Boolean)? = null): List<VMWord>? {
+            if (opcodes.size > (source.size - pc)) return null
+            var hit = true
+            var nulls = mutableListOf<VMWord>()
+            opcodes.forEachIndexed { i, t ->
+                if (t == null) nulls.add(source[pc + i])
+                else if ((pc + i) in jumpMap.keys) hit = false  // Miss if we overlap a jump dest
+                else if (source[pc + i].opcode != t) hit = false
+            }
+            if (hit && (check?.invoke(nulls) != false)) {
+                pc += opcodes.size
+                lastMatchSize = opcodes.size
```

Comment:
I have to save this lastMatchSize so I can subtract it further down when I'm finding the source VMWord to copy the tracing line+char data from.